include/envoy/http/header_map.h

```diff
@@ -625,42 +613,170 @@ class HeaderMap {
 
 using HeaderMapPtr = std::unique_ptr<HeaderMap>;
 
+/**
+ * Registry for custom headers. Headers can be registered multiple times in independent
+ * compilation units and will still point to the same slot. Headers are registered independently
+ * for each concrete header map type and do not overlap. Handles are strongly typed and do not
+ * allow mixing.
+ */
+class CustomInlineHeaderRegistry {
+public:
+  enum class Type { RequestHeaders, RequestTrailers, ResponseHeaders, ResponseTrailers };
+  using RegistrationMap = std::map<LowerCaseString, size_t>;
+
+  // A "phantom" type is used here to force the compiler to verify that handles are not mixed
+  // between concrete header map types.
+  template <Type type> struct Handle {
+    Handle(RegistrationMap::const_iterator it) : it_(it) {}
+
+    RegistrationMap::const_iterator it_;
+  };
+
+  /**
+   * Register an inline header and return a handle for use in inline header calls. Must be called
+   * prior to finalize().
+   */
+  template <Type type>
+  static Handle<type> registerInlineHeader(const LowerCaseString& header_name) {
+    static size_t inline_header_index = 0;
+
+    ASSERT(!mutableFinalized<type>());
+    auto& map = mutableRegistrationMap<type>();
+    auto entry = map.find(header_name);
+    if (entry == map.end()) {
```

Comment:
you can do this with one map lookup:
```
  auto insertion = map.insert(RegistrationMap::value_type(header_name, 0));
  if (insertion.second) {
    insertion.first.second = inline_header_index++;
  }
```
not sure if that's worth it as it might be less readable than what you have and mostly I'd guess we wouldn't have two filters registering the same header? Up to you.